sympy/polys/matrices/ddm.py

```diff
@@ -958,6 +958,83 @@ def lu(a):
 
         return L, U, swaps
 
+    def PLDUdecompositionFF(self):
+        """
+        Compute a fraction-free PLDU decomposition for DDM.
+
+        Returns
+        =======
+
+        (P, L, D, U)
+            P is the permutation matrix.
+            L is the lower triangular matrix.
+            D is the diagonal matrix.
+            U is the upper triangular matrix.
+
+        Raises
+        ======
+
+        DMRankError
+            If the matrix is not full rank.
+
+        See Also
+        ========
+
+        sympy.matrices.matrixbase.MatrixBase.LUdecomposition
+        LUdecomposition_Simple
+        LUsolve
+
+        References
+        ==========
+
+        .. [1] W. Zhou & D.J. Jeffrey, "Fraction-free matrix factors: new forms
+            for LU and QR factors". Frontiers in Computer Science in China,
+            Vol 2, no. 1, pp. 67-80, 2008.
+        """
+        rows, cols = self.shape
+        K = self.domain
+        P = self.eye(rows, K)
+        L = self.eye(rows, K)
+        D = self.zeros((rows, rows), K)
+        U = self.copy()
+
+        if rows == 1 and cols == 1:
+            D[0][0] = U[0][0]
+            U[0][0] = K.one
+            return P, L, D, U
+
+        oldpivot = K.one
+
+        for k in range(rows - 1):
+            if U[k][k] == K.zero:
+                for kpivot in range(k + 1, rows):
+                    if U[kpivot][k] != K.zero:
+                        U[k], U[kpivot] = U[kpivot], U[k]
+                        P[k], P[kpivot] = P[kpivot], P[k]
+                        L[k][:k], L[kpivot][:k] = L[kpivot][:k], L[k][:k]
+                        break
+                else:
+                    raise DMRankError("Matrix is not full rank")
+
+            Ukk = U[k][k]
+            L[k][k] = Ukk
+            D[k][k] = oldpivot * Ukk
+
+            for i in range(k + 1, rows):
+                Uik = U[i][k]
+                L[i][k] = Uik
+
+                for j in range(k + 1, cols):
+                    U[i][j] = (Ukk * U[i][j] - U[k][j] * Uik) / oldpivot
```

Comment:
True division with `/` should not be used unless the domain is a field. Instead `K.exquo` should be used for exact division:
```python
In [1]: ZZ(2) / ZZ(3)
Out[1]: 0.6666666666666666

In [2]: ZZ.exquo(ZZ(2), ZZ(3))
...
ExactQuotientFailed: 3 does not divide 2 in ZZ

In [3]: ZZ.exquo(ZZ(6), ZZ(3))
Out[3]: 2
```